### PR TITLE
Lagrange

### PR DESCRIPTION
**Pull request by swasilyev**

### CODE REVIEW

File: src/pcs/kzg/lagrange.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+impl<G: AffineRepr, D: EvaluationDomain<G::ScalarField>> LagrangianCK<G, D> {
+    pub fn generate<R: RngCore>(domain: D, rng: &mut R) -> Self {
+        let tau = G::ScalarField::rand(rng);
+        let g = G::Group::rand(rng);
+        Self::from_trapdoor(domain, tau, g)
+    }
+
+    pub fn from_trapdoor(domain: D, tau: G::ScalarField, g: G::Group) -> Self {
```
Comment (InaOana) [inline src/pcs/kzg/lagrange.rs:34]: Usually `tau` is considered/represents toxic waste thus is unknown to any party. Thus, I have not expected it to be an input to any function. Is this function meant to be used only in tests? If so, can it be defined only in tests or can it be defined as private?

File: src/pcs/kzg/lagrange.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        assert!(!domain.evaluate_vanishing_polynomial(tau).is_zero()); // doesn't give a basis
+        let lis_at_tau = domain.evaluate_all_lagrange_coefficients(tau); // L_i(tau)
+        let lis_in_g = single_base_msm(&lis_at_tau, g); // L_i(tau).G
+        Self { lis_in_g, domain }
+    }
+}
+
+impl<G: AffineRepr> MonomialCK<G> {
```
Comment (InaOana) [inline src/pcs/kzg/lagrange.rs:42]: nit1: Would it make sense to move this implementation to the place where `MonomialCK` is defined?
nit2: Would it make more sense to rename `MonomialCK` to `MonomialSRS`?

File: src/pcs/kzg/lagrange.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+use ark_std::rand::RngCore;
+use ark_std::UniformRand;
+
+use crate::pcs::CommitterKey;
+use crate::pcs::kzg::params::MonomialCK;
+use crate::utils::ec::single_base_msm;
+
+/// Used to commit to univariate polynomials represented in the evaluation form.
+#[derive(Clone, Debug, CanonicalSerialize, CanonicalDeserialize)]
+pub struct LagrangianCK<G: AffineRepr, D: EvaluationDomain<G::ScalarField> = GeneralEvaluationDomain<<G as AffineRepr>::ScalarField>> {
```
Comment (InaOana) [inline src/pcs/kzg/lagrange.rs:14]: nit: Would it make more sense to rename `LagrangianCK` to `LagrangianSRS`?

File: src/pcs/kzg/mod.rs [hunk 4]
```diff
@@ -92,32 +93,39 @@ impl<E: Pairing> KZG<E> {
         let acc_opening = Self::accumulate(openings, &coeffs, vk);
         Self::verify_accumulated(acc_opening, vk)
     }
+
+    fn _commit(coeffs: &[E::ScalarField], bases: &[E::G1Affine]) -> KzgCommitment<E> {
```
Comment (InaOana) [inline src/pcs/kzg/mod.rs:97]: nit: Would it make more sense to rename `_commit` to `new` or `from_...`?

File: src/pcs/kzg/params.rs [hunk 5]
```diff
@@ -1,23 +1,36 @@
...
+    fn ck(&self) -> Self::CK {
+        let monomial = MonomialCK {
+            powers_in_g1: self.powers_in_g1.clone()
+        };
+        KzgComitterKey { monomial, lagrangian: None }
+    }
+
+    fn ck_with_lagrangian(&self, domain_size: usize) -> Self::CK {
+        let domain = GeneralEvaluationDomain::new(domain_size).unwrap();
+        assert_eq!(domain.size(), domain_size, "domains of size {} are not supported", domain_size);
```
Comment (InaOana) [inline src/pcs/kzg/params.rs:24]: Is it possible for `GeneralEvaluationDomain::new` to generate a domain of a different size than the expected `domain_size`?

File: src/pcs/kzg/params.rs [hunk 6]
```diff
@@ -1,23 +1,36 @@
...
+        assert!(domain_size <= self.powers_in_g1.len());
+        let monomial = MonomialCK {
+            powers_in_g1: self.powers_in_g1[0..domain_size].to_vec()
```
Comment (InaOana) [inline src/pcs/kzg/params.rs:27]: nit: 0 can be removed from `[0..domain_size]` to `[..domain_size]`

File: src/pcs/kzg/params.rs [hunk 7]
```diff
@@ -93,11 +118,10 @@ pub struct KzgVerifierKey<E: Pairing> {
 
 impl<E: Pairing> VerifierKey for KzgVerifierKey<E> {}
 
-impl<E: Pairing> From<KzgVerifierKey<E>> for MonomialCK<E::G1Affine> {
+impl<E: Pairing> From<KzgVerifierKey<E>> for KzgComitterKey<E::G1Affine> {
     fn from(vk: KzgVerifierKey<E>) -> Self {
-        MonomialCK {
-            powers_in_g1: vec![vk.g1]
-        }
+        let monomial = MonomialCK { powers_in_g1: vec![vk.g1] };
```
Comment (InaOana) [inline src/pcs/kzg/params.rs:123]: I am not sure if I understand this logic. If I understood correctly, the powers in `g_1` for `MonomialCK` should include more than one value.

File: src/pcs/kzg/lagrange.rs (hunk 1 above)
Comment (swasilyev) [reply src/pcs/kzg/lagrange.rs:34]: The only secure way to generate params for KZG is an MPC. So ideally it should be in #[cfg(test)], but there's still https://github.com/w3f/fflonk/blob/c10a95bd9a4be8025a9cc961a34969b0752de687/src/pcs/mod.rs#L81 on the trait that might make sense for other schemes.

File: src/pcs/kzg/lagrange.rs (hunk 2 above)
Comment (swasilyev) [reply src/pcs/kzg/lagrange.rs:42]: 1. That's fine either way. Depends on if i decide to implement more lagrangian tricks, like openning in a root.
2. I think it's a ck. Srs is always generated monomial, and should contain at least tauG2 to be useful.

File: src/pcs/kzg/params.rs (hunk 5 above)
Comment (swasilyev) [reply src/pcs/kzg/params.rs:24]: Absolutely. Basically it generates you a domain for the next power of 2.

File: src/pcs/kzg/params.rs (hunk 6 above)
Comment (swasilyev) [reply src/pcs/kzg/params.rs:27]: sure

File: src/pcs/kzg/params.rs (hunk 7 above)
Comment (swasilyev) [reply src/pcs/kzg/params.rs:123]: It's an ugly bit. See https://github.com/w3f/fflonk/blob/271ed0e776f6444375ac891f1e1a4a43ef25eb9a/src/pcs/mod.rs#L84 Will change some day.

**Pull request merged**